exploration/variable-mutability.md

```diff
@@ -0,0 +1,244 @@
+  > ```
+  > let $foo = {$bar :uppercase}
+  > let $baz = {$someNumber :number groupingUsed=false}
+  > ```
+- Users, such as translators, want to annotate a variable
+  (either local or external) without invalidating
+  existing use of the variable in pattern strings. For example:
+  > ```
+  > let $foo = {$foo :transform}
+  > match {$a :plural} {$b :plural}
+  > when 0   0   {...{$foo}...}
+  > when 0   one {...{$foo}...}
+  > when 0   *   {...{$foo}...}
+  > when one 0   {...{$foo}...}
+  > when one one {...{$foo}...}
+  > when one *   {...{$foo}...}
+  > when *   0   {...{$foo}...}
+  > when *   one {...{$foo}...}
+  > when *   *   {...{$foo}...}
+  > ```
+- Users want to perform multiple transforms on a value.
+  Since our syntax does not permit embedding or chaining, this requires multiple declarations.
+  > ```
+  > let $foo = {$foo :text-transform transform=uppercase}
+  > let $foo = {$foo :trim}
+  > let $foo = {$foo :sanitize target=html}
+  > ```
+  >
+  > This can also be achieved by renaming:
+  >
+  > ```
+  > let $foo1 = {$foo :text-transform transform=uppercase}
+  > let $foo2 = {$foo1 :trim}
+  > let $foo3 = {$foo2 :sanitize target=html}
+  > ```
+- Users want to impose typing on (we say "annotate") external variables
+  or literals:
+  > ```
+  > let $fooAsNumber = {$foo :number}
+  > let $anotherNumber = {42 :number}
+  > ```
+- Users may wish to provide complex annotations which are reused across mulitple patterns
+
+  > ```
+  > let $count = {$count :number}
+  > let $date = {$date :datetime dateStyle=long}
+  > match {$count}
+  > when 1 {You received one message on {$date}}
+  > when * {You received {$count} messages on {$date}}
+  > ```
+
+- Implementers need to know what value is associated with a named variable, see #299.
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+- Be able to re-annotate variables without having to rename them in the message body
+- Allow static analysis to detect mistakes when referencing an undefined local variable
+- Be able to re-annotate variables multiple times (because we do not allow nesting)
+
+- _more needed?_
+
+## Constraints
+
+_What prior decisions and existing conditions limit the possible design?_
+
+- Variable names are potentially contrained by `Nmtoken`.
+  The reason we chose Nmtoken (and Name) was to maximize compatibility with (potential)
+  LDML constructs, since CLDR uses XML.
+  The "carve out" for various sigils doesn't conflict with naming in LDML currently and
+  future conflicts are under our (CLDR-TC's) control.
+
+## Proposed Design
+
+_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+
+Separate local variables from externally passed values by altering the sigil
+and by using a visually distinctive pattern for local names
+(in an effort to prevent `$foo`/`@foo` confusion).
+
+```abnf
+variable     = local_var / external_var
+local_var    = "@_" name
```

Comment:
Yeah, no. We categorically do not need any two-character prefixes.

Reply:
See later discussion. This is horrifying while we agree on something...